larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -68,123 +69,77 @@ DenseElementsAttr Bitpack(PatternRewriter& builder, Attribute x) {
-    }
+  const auto channels = existing_shape[3];
+  const auto packed_channels = (channels + 32 - 1) / 32;
+  return RankedTensorType::get({existing_shape[0], existing_shape[1],
+                                existing_shape[2], packed_channels},
+                               rewriter.getIntegerType(32));
+}
 
-    constexpr int bitwidth = 32;  // We use 32-bit bitpacking.
+template <typename BinaryOp>
+struct SetBitpackedActivations : public OpRewritePattern<BinaryOp> {
+  using OpRewritePattern<BinaryOp>::OpRewritePattern;
 
-    /*********************
-     * Compute new types *
-     *********************/
+  LogicalResult matchAndRewrite(BinaryOp outer_binary_op,
+                                PatternRewriter& rewriter) const override {
+    Operation* input_op = outer_binary_op.input().getDefiningOp();
 
-    RankedTensorType new_inner_bconv_op_type;
-    RankedTensorType new_intermediate_maxpool_op_type;
+    // If the input op has more than one use, we can't apply an optimisation.
+    if (!input_op || !input_op->hasOneUse()) return failure();
 
-    // If applicable, compute the new type for the inner bconv op.
+    // Try and match `input_op` to a binary convolution.
+    auto inner_bconv_op = dyn_cast_or_null<TF::LceBconv2dOp>(input_op);
     if (inner_bconv_op) {
-      ShapedType inner_bconv_op_type =
-          inner_bconv_op.getType().cast<ShapedType>();
-
-      // Don't apply this transformation if the type is already I32.
-      if (inner_bconv_op_type.getElementType().isInteger(bitwidth)) {
+      if (inner_bconv_op.padding() == "SAME" &&
+          inner_bconv_op.pad_values() != 1) {
         return failure();
       }
 
-      // Compute the new type for the inner bconv op.
-      const auto inner_bconv_op_shape = inner_bconv_op_type.getShape();
-      if (inner_bconv_op_shape.size() != 4) return failure();
-      const auto channels = inner_bconv_op_shape[3];
-      const auto packed_channels = (channels + bitwidth - 1) / bitwidth;
-      new_inner_bconv_op_type = RankedTensorType::get(
-          {inner_bconv_op_shape[0], inner_bconv_op_shape[1],
-           inner_bconv_op_shape[2], packed_channels},
-          rewriter.getIntegerType(bitwidth));
-    }
+      auto maybe_bitpacked_type = getBitpackedType(
+          rewriter, inner_bconv_op.getType().cast<ShapedType>());
+      if (!maybe_bitpacked_type.hasValue()) return failure();
 
-    // If applicable, compute the new type for the intermediate maxpool op.
-    if (intermediate_maxpool_op) {
-      ShapedType intermediate_maxpool_op_type =
-          intermediate_maxpool_op.getType().cast<ShapedType>();
+      rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
+          inner_bconv_op, maybe_bitpacked_type.getValue(),
+          inner_bconv_op.getOperands(), inner_bconv_op.getAttrs());
 
-      // Don't apply this transformation if the type is already I32.
-      if (intermediate_maxpool_op_type.getElementType().isInteger(bitwidth)) {
+      return success();
```

Comment:
I had hoped that we can use something like the following here so we don't need to call `hasValue` and `getValue`, if this this doesn't work, I'm happy with the current solution.

```cpp
if (auto bitpacked_type = getBitpackedType(rewriter, inner_bconv_op.getType().cast<ShapedType>())) {
  ...
  return success();
}
return failure();

Reply:
When using this `bitpacked_type` is still the optional type, so you have to call `getValue()` inside the if block.

Given that's the case, I think I'd rather be explicit and use the `hasValue()` check rather than relying on the truthiness of `llvm::None` being false.

Reply:
Makes sense, I wasn't aware of that.

Though judging from how it is [used in TensorFlow](https://github.com/tensorflow/tensorflow/blob/e164659e5b607f2e7fc54fbc090894c78745d544/tensorflow/compiler/mlir/lite/flatbuffer_export.cc#L1211-L1213) one should be able to do:
```cpp
if (auto bitpacked_type = getBitpackedType(rewriter, inner_bconv_op.getType().cast<ShapedType>())) {
  rewriter.replaceOpWithNewOp<TF::LceBconv2dOp>(
          inner_bconv_op, *bitpacked_type,
          inner_bconv_op.getOperands(), inner_bconv_op.getAttrs());
  return success();
}
```
If you pass it using `*bitpacked_type` it might work, but I migth miss something. In both cases this PR looks good to me :+1:

Reply:
Nice, you're right it does work with the dereference.